app/apollo/resolvers/subscription.js

```diff
@@ -672,26 +725,36 @@ const subscriptionResolvers = {
         */
 
         const conditions = await getGroupConditionsIncludingEmpty(me, org_id, ACTIONS.READ, 'name', queryName, context);
-        logger.debug({req_id, user, org_id, conditions }, `${queryName} group conditions are...`);
-        var subscription = await models.Subscription.findOne({ org_id, uuid, ...conditions }, {}).lean({ virtuals: true });
+        logger.debug({req_id, user, org_id, conditions}, `${queryName} group conditions are...`);
 
-        if(!subscription){
-          throw new NotFoundError(context.req.t('Subscription { uuid: "{{uuid}}", org_id:{{org_id}} } not found.', {'uuid':uuid, 'org_id':org_id}), context);
-        }
+        // Find the subscription
+        const subscription = await models.Subscription.findOne({ org_id, uuid, ...conditions }, {}).lean({ virtuals: true });
+        logger.info({req_id, user, org_id, uuid, version_uuid}, `${queryName} validating - found: ${!!subscription}`);
 
-        // this may be overkill, but will check for strings first, then groups below
-        await validAuth(me, org_id, ACTIONS.SETVERSION, TYPES.SUBSCRIPTION, queryName, context, [subscription.uuid, subscription.name]);
+        const identifiers = subscription ? [uuid, subscription.name] : [uuid];
+        await validAuth(me, org_id, ACTIONS.SETVERSION, TYPES.SUBSCRIPTION, queryName, context, identifiers);
+        logger.info( {req_id, user, org_id, uuid, version_uuid}, `${queryName} validating - authorized` );
 
         // validate user has enough cluster groups permissions to for this sub
         // TODO: we should use specific groups action below instead of manage, e.g. setSubscription action
         const allowedGroups = await getAllowedGroups(me, org_id, ACTIONS.SETVERSION, 'name', queryName, context);
+
+        // Find the channel
+        const channel = await models.Channel.findOne({ org_id, uuid: subscription.channel_uuid });
+        logger.info({req_id, user, org_id, uuid, version_uuid}, `${queryName} validating - found: ${!!channel}`);
+
+        const channelIdentifiers = channel ? [subscription.channel_uuid, channel.name] : [subscription.channel_uuid];
+        await validAuth(me, org_id, ACTIONS.READ, TYPES.CHANNEL, queryName, context, channelIdentifiers);
+        logger.info({req_id, user, org_id, uuid, version_uuid}, `${queryName} validating - channel authorized`);
+
+        // Check for errors after validation
+        if (!subscription) {
+          throw new NotFoundError(context.req.t('Subscription { uuid: "{{uuid}}", org_id:{{org_id}} } not found.', {'uuid':uuid, 'org_id':org_id}), context);
+        }
```

Comment:
- Validation that the Subscription is found is _part_ of validation, not "after validation".
- This code block should be moved up so that it is checked immediately verifying auth to the subscription name/uuid, before querying the Channel or verifying auth on it -- if the subscription doesnt even exist, there's no reason to do the expensive channel verification.

Reply:
Similarly the below `if (subscription.groups.some(.... ` block should also be checked before bothering with the expensive channel lookup and auth check.

Moving these two blocks to the logical spot would result in the `if(!channel){` block being co-located with the block where the `channel` variable is found and auth checked, which makes the code easier to read.

Reply:
Resolved